execute/exectypes/costly_messages.go

```diff
@@ -0,0 +1,135 @@
+	// Also takes in a map from message ID to the message's timestamp, needed to calculate fee boosting.
+	Observe(
+		ctx context.Context,
+		obs MessageObservations,
+		messageTimestamps map[cciptypes.Bytes32]time.Time,
+	) ([]cciptypes.Bytes32, error)
+}
+
+func NewCostlyMessageObserver() CostlyMessageObserver {
+	return &CcipCostlyMessageObserver{
+		// TODO: Implement fee and exec cost calculators
+		feeCalculator:      &NoOpMessageFeeE18USDCalculator{},
+		execCostCalculator: &NoOpMessageExecCostE18USDCalculator{},
+	}
+}
+
+type CcipCostlyMessageObserver struct {
+	feeCalculator      MessageFeeE18USDCalculator
+	execCostCalculator MessageExecCostE18USDCalculator
+}
+
+// Observe returns a slice of message IDs that are too costly to execute.
+// It calculates the fee and execution cost of each message. The messages are considered too costly if the fee is less
+// than the execution cost.
+func (o *CcipCostlyMessageObserver) Observe(
+	ctx context.Context,
+	obs MessageObservations,
+	messageTimestamps map[cciptypes.Bytes32]time.Time,
+) ([]cciptypes.Bytes32, error) {
+	messages := make([]cciptypes.Message, 0)
+	for _, seqNumToMsg := range obs {
+		for _, msg := range seqNumToMsg {
+			messages = append(messages, msg)
+		}
+	}
+
+	messageFees, err := o.feeCalculator.MessageFeeE18USD(ctx, messages, messageTimestamps)
+	if err != nil {
+		return nil, fmt.Errorf("unable to calculate message fees: %w", err)
+	}
+
+	execCosts, err := o.execCostCalculator.MessageExecCostE18USD(ctx, messages)
+	if err != nil {
+		return nil, fmt.Errorf("unable to calculate message execution costs: %w", err)
+	}
+
+	costlyMessages := make([]cciptypes.Bytes32, 0)
+	for _, msg := range messages {
+		fee, ok := messageFees[msg.Header.MessageID]
+		if !ok {
+			return nil, fmt.Errorf("missing fee for message %s", msg.Header.MessageID)
+		}
+		execCost, ok := execCosts[msg.Header.MessageID]
+		if !ok {
+			return nil, fmt.Errorf("missing exec cost for message %s", msg.Header.MessageID)
+		}
+		if fee.Cmp(execCost.Int) < 0 {
+			costlyMessages = append(costlyMessages, msg.Header.MessageID)
+		}
+	}
+
+	return costlyMessages, nil
+}
+
+var _ CostlyMessageObserver = &CcipCostlyMessageObserver{}
+
+// MessageFeeE18USDCalculator Calculates the fees (paid at source) of a set of messages in e-18 USDs.
+type MessageFeeE18USDCalculator interface {
+	// MessageFeeE18USD Returns a map from message ID to the message's fee in e-18 USDs. For example, if the message's
+	// fee is 12USD, this function return this message's fee as 12 * 1e18. You can think of this function returning the
+	// fee not in USD, but in a small denomination of USD, analogous to returning the cost in wei instead of ETH
+	// (1 wei = 1e-18 ETH).
+	MessageFeeE18USD(
+		ctx context.Context,
+		messages []cciptypes.Message,
+		messageTimestamps map[cciptypes.Bytes32]time.Time,
+	) (map[cciptypes.Bytes32]cciptypes.BigInt, error)
```

Comment:
Taking a stab at designing the types for this to ensure the denominations are always honored:

```golang
type USD18 *big.Int

func NewUSD18FromUSD(usd uint64) USD18 {
  return new(big.Int).Mul(new(big.Int).SetUint64(usd), big.NewInt(1e18))
}

// ... operations ...
```

Maybe this'll become clearer when we see what kind of operations we do on these values, including whats in gas prices and token prices. cc @asoliman92